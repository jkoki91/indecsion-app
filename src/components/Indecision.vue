<template>
  <img v-if="img" :src="img" alt="bg" />
  <div class="bg-dark"></div>

  <div class="indecision-container">
    <input 
    type="text" 
    placeholder="Hazme una pregunta" 
    v-model="question"
    />
    <!-- v-on:keyup.enter="question" esto no se si vale para algo -->
    <p>Recuerda terminar con un signo de interrogación (?)</p>

    <div>
        <!-- <h2>Seré millonario??</h2>  -->
        <h2>{{question || 'Seré millonario??'}}</h2> 
        <h1>{{answer || 'Si, No, ... pensando'}}</h1>
    </div>
  </div>
</template>

<script>

export default {
    data(){
        return{
            question: null,
            answer: null,
            img: null,
            isValidQuestion: false
        }
    },
    watch:{
        question( value,oldValue ){
            this.isValidQuestion = false

            if(!value.includes('?')) return

            this.isValidQuestion = true

            this.getAnswer()
        }
    },
    methods:{
        async getAnswer() {
            this.answer = 'Pensando...'

            const { answer,image} = await fetch('https://yesno.wtf/api').then(r=>r.json())
            // .then(d=>console.log(d))
            if (answer==='yes'){
                this.answer = 'SI'
            }else{
                if(answer==='no'){
                this.answer = 'NO!!!'
                }else{this.answer = 'TAL VEZ' }
            }
            // this.answer = answer
            this.img = image

            // console.log(data)
        }
    }
};  
</script>

<style scoped>
img,
.bg-dark {
  height: 100vh;
  left: 0px;
  max-height: 100%;
  max-width: 100%;
  position: fixed;
  top: 0px;
  width: 100vw;
}

.bg-dark {
  background-color: rgba(0, 0, 0, 0.4);
}

.indecision-container {
  position: relative;
  z-index: 99;
}

input {
  width: 250px;
  padding: 10px 15px;
  border-radius: 5px;
  border: none;
}
input:focus {
  outline: none;
}

p {
  color: white;
  font-size: 20px;
  margin-top: 0px;
}

h1,
h2 {
  color: white;
}

h2 {
  margin-top: 150px;
}
</style>