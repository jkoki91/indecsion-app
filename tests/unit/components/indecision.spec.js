import { shallowMount } from '@vue/test-utils'
import Indecision from '@/components/Indecision'


describe('Indecision Component', () => {
    let wrapper
    let clgSpy

    global.fetch = jest.fn(() => Promise.resolve({
        json: ()=> Promise.resolve({
            answer: 'yes',
            forced: false,
            image: 'https://yesno.wtf/assets/yes/2.gif'
        })
    }))

    beforeEach(()=>{
        wrapper = shallowMount( Indecision )

        clgSpy = jest.spyOn( console, 'log' )

        jest.clearAllMocks()
    })

    / test('debe hacer match con el snapshot',()=>{

        expect(wrapper.html()).toMatchSnapshot()
    
    })

    test('escribir en el input no debe de disparar nada (console.log)', async()=>{

        // console.log(wrapper)
        // console.log(wrapper.vm)

        const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')
        
        const input = wrapper.find('input')
        await input.setValue('Hola Mundo')

        expect( clgSpy ).toHaveBeenCalled()
        // expect( clgSpy ).toHaveBeenCalledTimes(2)
        // expect( getAnswerSpy ).toHaveReturnedTimes(0)
        expect( getAnswerSpy ).not.toHaveBeenCalled()

    })

    test('escribir el simbolo de interrogaciÃ³n debe disparar el getAnswer', async() => { 
        console.log(wrapper.vm)

        const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')

        const input = wrapper.find('input')
        await input.setValue('Hola Mundo?')

        // expect( clgSpy ).toHaveBeenCalledTimes(2)
        expect( clgSpy ).toHaveBeenCalled()
        expect( getAnswerSpy ).toHaveBeenCalled()


    })

    test('pruebas en get answer', async() => { 

        await wrapper.vm.getAnswer()

        // console.log(wrapper.vm.img)
        // console.log(wrapper.vm.answer)

        const img = wrapper.find('img')

        expect( img.exists() ).toBeTruthy()
        expect( wrapper.vm.img ).toBe('https://yesno.wtf/assets/yes/2.gif')
        expect( wrapper.vm.answer ).toBe('SI')

    })

    test('pruebas en get answer - Fallo API', async() => { 

        fetch.mockImplementationOnce( () => Promise.reject('API is down') )

        await wrapper.vm.getAnswer()

        const img = wrapper.find('img')

        expect( img.exists() ).toBeFalsy()
        expect( wrapper.vm.answer ).toBe('No se pudo cargar el API')

    })

})