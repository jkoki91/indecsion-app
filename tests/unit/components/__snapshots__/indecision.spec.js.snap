// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision Component debe hacer match con el snapshot 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" placeholder="Hazme una pregunta"><!-- v-on:keyup.enter="question" esto no se si vale para algo -->
  <p>Recuerda terminar con un signo de interrogación (?)</p>
  <div>
    <!-- <h2>Seré millonario??</h2>  -->
    <h2>Seré millonario??</h2>
    <h1>Si, No, ... pensando</h1>
  </div>
</div>
`;
